plutus-core/untyped-plutus-core/test/Analysis/Spec.hs

```diff
@@ -1,66 +1,148 @@
+  we have the following patterns:
+
+  - Builtin functions with 0 type arguments followed by N>0 term arguments:
+      - EncodeUtf8 : [ string ] -> bytestring
+      - AddInteger : [ integer, integer ] -> integer
+      - WriteBits  : [ bytestring, list(integer), bool ] -> bytestring
+
+  - Builtin functions with 1 type argument followed by N>0 term arguments:
+      - HeadList   : [ forall a, list(a) ] -> a
+      - Trace      : [ forall a, string, a ] -> a
+      - IfThenElse : [ forall a, bool, a, a ] -> a
+
+  - Builtin functions with 2 type arguments followed by N>0 term arguments:
+      - FstPair    : [ forall a, forall b, pair(a, b) ] -> a
+      - ChooseList : [ forall a, forall b, list(a), b, b ] -> b
+      - CaseList   : [ forall a, forall b
+                      , b
+                      , (fun a (fun [ (con list) a ] b))
+                      , list(a)
+                      ] -> b
+
+  When it comes to purity of builtin terms,
+  we want to test the following cases:
+  +---------------+---------------+
+  | Type args     | Term args     |
+  | left to apply | left to apply |
+  +---------------+---------------+
+  |     some      |     some      |
+  |     none      |     some      |
+  |     none      |     none      |
+  +---------------+---------------+
+-}
+
+testSomeTypeSomeTermArgsLeft :: TestTree
+testSomeTypeSomeTermArgsLeft =
+  testCase "some type args and some term args are unapplied" $
+    map (isPure builtinSemantics) terms @?= [True, True]
+ where
+  terms :: [Term Name DefaultUni DefaultFun ()] =
+    [ Builtin () Trace -- 1 type arg and 2 term args are unapplied
+    , Force () (Builtin () FstPair) -- 2 type args, 1 applied and 1 left
+    ]
+
+testNoTypeSomeTermArgsLeft :: TestTree
+testNoTypeSomeTermArgsLeft =
+  testCase "no type args and some term args are unapplied" $
+    map (isPure builtinSemantics) terms @?= [True, True]
+ where
+  terms :: [Term Name DefaultUni DefaultFun ()] =
+    [ Builtin () EncodeUtf8 -- no type args, 1 term arg left to apply
+    , Force () (Builtin () Trace) -- 1 type arg applied, 2 term args left
+    ]
+
+testNoTypeNoTermArgsLeft :: TestTree
+testNoTypeNoTermArgsLeft =
+  testGroup
+    "no type args and no term args are unapplied"
+    [ testAddInteger
+    , testIfThenElse
+    ]
+ where
+  testAddInteger :: TestTree =
+    testCase "AddInteger" $ isPure builtinSemantics term @?= False
+   where
+    term :: Term Name DefaultUni DefaultFun () =
+      Apply () (Apply () (Builtin () AddInteger) (termVar 1)) (termVar 2)
+
+  testIfThenElse :: TestTree =
+    testCase "IfThenElseApplied" $
+      isPure builtinSemantics termIfThenElse @?= False
+
+testForceNoTypeParam :: TestTree
+testForceNoTypeParam =
+  testCase "forcing a non-existing type param is impure" $
+    isPure builtinSemantics (Force () (Builtin () EncodeUtf8)) @?= False
+
+testApplyNoTermParam :: TestTree
+testApplyNoTermParam =
+  testCase "applying a non-existing term param is impure" $
```

Comment:
There are two subcases here:
- Applying a term param, but a type param is expected.
- Applying a term param, but the builtin has already been saturated (i.e., over-saturated builtin)

The first one is definitely an error, but the second one could be valid, e.g., `ifThenElse True (\i -> i + 1) (\i -> i + 2) 42`. They are both impure, but for different reasons.